Generate ECS detection rules

leehinman opened the pull request:
- generates required rules
 - generates allowed_values rules
 - generates expected_values rules for event.category

## Example Required rule

```
{
  "description": "This rule checks to make sure the required field `@timestamp` is present.",
  "from": "now-1d",
  "interval": "1d",
  "name": "ECS Check: @timestamp required field",
  "query": "not @timestamp:*",
  "risk_score": 42,
  "severity": "high",
  "type": "query"
}
```

## Example allowed_values rule

```
{
  "description": "This rule checks to make sure the `event.outcome` field only contains allowed values.",
  "from": "now-1d",
  "interval": "1d",
  "name": "ECS Check: event.outcome allowed values",
  "query": "event.outcome:* and not event.outcome:failure and not event.outcome:success and not event.outcome:unknown",
  "risk_score": 42,
  "severity": "high",
  "type": "query"
}
```

## Example expected_values rule

```
{
  "description": "This rule checks to make sure that `event.type` contains expected values when `event.category` is web.",
  "from": "now-1d",
  "interval": "1d",
  "name": "ECS Check: expected `event.type` for `event.category` web",
  "query": "event.category:web and event.type:* and not event.type:access and not event.type:error and not event.type:info",
  "risk_score": 42,
  "severity": "high",
  "type": "query"
}
```

## Added Gherkin feature files.

With a feature file that looks like this:
``` gherkin
Feature: Address field should be populated if ip or domain is polulated.
  Scenario: If the source.ip field exists then so should the source.address field.
    Given an event
    And the source.ip field is present
    Then the source.address field is present
```

We get documentation that looks like:
![Screen Shot 2021-04-30 at 12 35 36](https://user-images.githubusercontent.com/57081003/116732691-c63a9b80-a9b0-11eb-816a-82331dc66136.png)

And a detection rule that looks like:
``` json
{
  "description": "If the source.ip field exists then so should the source.address field.",
  "from": "now-1d",
  "interval": "1d",
  "name": "ECS Check: If the source.ip field exists then so should the source.address field.",
  "query": "source.ip:* and not source.address:*",
  "risk_score": 42,
  "severity": "high",
  "type": "query"
}
```

ebeahan commented on scripts/generators/rules_generator.py:
```diff
@@ -0,0 +1,81 @@
+    for field_name, field_properties in required_fields.items():
+        if field_properties['required'] == False:
+            next
+        rule = {"name": "ECS Check: {0} required field".format(field_name),
```
A possible change for later:

Similar how some of the generated asciidoc files use [templates](https://github.com/elastic/ecs/tree/master/scripts/templates), we could consider creating rule templates to have some of the rule configuration outside the rule generator here.

ebeahan commented on scripts/generator.py:
```diff
@@ -59,6 +60,7 @@ def main():
     es_template.generate(nested, ecs_version, out_dir, args.mapping_settings)
     es_template.generate_legacy(flat, ecs_version, out_dir, args.template_settings, args.mapping_settings)
     beats.generate(nested, ecs_version, out_dir)
+    rules_generator.generate(flat, ecs_version, out_dir)
```
Will we want to include the also add the generated rules file? Not seeing the `generated/rules/` directory included right now.

ebeahan commented on scripts/generators/rules_generator.py:
```diff
@@ -0,0 +1,81 @@
+    return rules
+
+
+def save_rules(filename, rule_list):
```
Looks like there's also a similar function in the `es_template` generator. Might be a good candidate to move into `ecs_helpers`.

leehinman replied:
I ended up not using ecs_helpers.py but a new file because using ecs_helper.py with behave was giving me import errors because import paths were different.  Probably worth fixing in a follow on PR.

ebeahan reviewed (commented):
Thanks, @leehinman! A few observations here and below.


* I think the types and content of the rules looks great.
* I noticed `rules.json` is a collection of JSON objects, but it isn't ndjson. Were you thinking the generated rules file would be something that could be uploaded as detection rules directly, or for another use?
* If we do add `generated/rules` as a maintained generated artifact, can we add an entry in the generated artifacts [README](https://github.com/elastic/ecs/blob/master/generated/README.md)?

leehinman commented:
> Thanks, @leehinman! A few observations here and below.
> 
>     * I think the types and content of the rules looks great.
> 
>     * I noticed `rules.json` is a collection of JSON objects, but it isn't ndjson. Were you thinking the generated rules file would be something that could be uploaded as detection rules directly, or for another use?

Thanks, fixed that.

> 
>     * If we do add `generated/rules` as a maintained generated artifact, can we add an entry in the generated artifacts [README](https://github.com/elastic/ecs/blob/master/generated/README.md)?

Thanks, added to generated/README.md

ebeahan commented:
@leehinman are you ready for another review, or is this still something you're iterating on?

leehinman commented:
> @leehinman are you ready for another review, or is this still something you're iterating on?

ready for another review

djptek commented:
@leehinman there's a few conflicts have emerged & `elasticsearch-ci/docs` may need a re-run

ebeahan reviewed (commented):
Sorry for not reviewing again sooner, @leehinman.

I like the use of Gherkin to specify the best practices as feature files. I think the plain-text spec works well to describe the outcomes we look for in ECS.

I also like the idea of having a true ECS `Best Practices` section in the docs, but I'm unsure if a collection of bullet points is the best way to communicate those concepts to users. However, maybe the bullet points can be high-level, brief descriptions and more detailed examples could follow later.

It took me a couple of passes to wrap my head around where the different ECS rule sets were coming from in the generator, but I think with some documentation that could help avoid user confusion.

If this is something we still want to try and introduce into the ECS tooling, we should also add some supporting unit testing.

ebeahan commented:
@leehinman Is this PR still something you want to move forward on?

ebeahan commented:
@leehinman I'm going to close this PR out for now. Feel free to re-open in the future.
